File: BST.java

public class BST {
	
	Node root;
	public class Node{
		int data;
		Node left;
		Node right;
		public Node(int data) {
			this.data = data;
			left = null;
			right = null;
		}
	}

	BST(){
		root = null;
	}
	
	void add(int num) {
		// if there is nothing in the node.
		root = addRecursion(root, num);
	}

	private Node addRecursion(Node newRoot, int num) {
		if(newRoot ==null) {
			newRoot = new Node(num);
			return newRoot;
		}
		
		else if(num < newRoot.data) {
			newRoot.left = addRecursion(newRoot.left, num);
		}
		else if(num > newRoot.data) {
			newRoot.right = addRecursion(newRoot.right, num);
		}
		return newRoot;
	}
	
	
	
	void preorder(Node root) {
		if(root!= null) {
			System.out.println(root.data);
			preorder(root.left);
			preorder(root.right);
		}
	}
	void inorder(Node root) {
		if(root != null) {
			inorder(root.left);
			System.out.println(root.data);
			inorder(root.right);
		}
	}
	void postorder(Node root) {
		if(root != null) {
			postorder(root.left);
			postorder(root.right);
			System.out.println(root.data);
		}
	}
	
	public static void main(String[] args) {
		BST a = new BST();
		a.add(10);
		a.add(1);
		a.add(5);
		a.add(20);
		a.add(30);
		a.add(8);
		a.add(25);
		System.out.println("INORDER : ");
		a.inorder(a.root);
		
		System.out.println();
		System.out.println("PREORDER : ");
		a.preorder(a.root);

		System.out.println();
		System.out.println("POSTORDER : ");
		a.postorder(a.root);
	}
	
	
}

File: AlgorithmInCSharp/JobInterviewPrepare/JobInterviewPrepare/BST/BST.cs

namespace JobInterviewPrepare
{
    public class BSTNode
    {
        public int val;
        public BSTNode left, right;
        public BSTNode(int val)
        {
            this.val = val;
        }
    }
    public class BST
    {
        public BSTNode Search(int value)
        {
            return null;
        }
        public BSTNode Insert(BSTNode root, int value)
        {
            if(root == null)
            {
                root = new BSTNode(value);
            }
            else if (value < root.val)
            {
                root.left = Insert(root.left, value);
            }else
            {
                root.right = Insert(root.right, value);
            }
            return root;
        }
        public void Traverse(BSTNode root)
        {
            if(root == null)
            {
                return;
            }
            System.Console.WriteLine($"The root {root.val}");
            Traverse(root.left);
            Traverse(root.right);
        }

        public BSTNode Find (BSTNode root, int value)
        {
            if (root == null) return null;
            if(root.val == value)
            {
                System.Console.WriteLine($"Fidn the value {root.val}");
                return root;
            }
            if(value  < root.val)
            {
                return Find(root.left, value);
            }
            else
            {
                return Find(root.right, value);
            }
        }
        public void PreOrder(BSTNode root)
        {

        }
        public void InOrder(BSTNode root)
        {
            if(root == null)
            {
                return;
            }
        }
        public void Remove(BSTNode root,int value){
            if (root == null) return;
        }
    }
    //iterative. 
    public class BSTree
    {
        public BSTNode Root { get; set; }
        public bool Insert(int value)
        {
            BSTNode before = null, after = this.Root; 
            while(after != null)
            {
                before = after;
                if(value < after.val)
                {
                    after = after.left;
                }
                else if( value > after.val)
                {
                    after = after.right;
                }
                else
                {
                    return false;
                }
            }
            BSTNode newNode = new BSTNode(value);
            if(this.Root == null)
            {
                this.Root = newNode;
            }
            else
            {
                if(value < before.val)
                {
                    before.left = newNode;
                }
                else
                {
                    before.right = newNode;
                }
            }
            return true;
        }

        public BSTNode Find(int value)
        {
            return this.Find(this.Root, value);
        }

        // Required to upgrade. 
        public BSTNode Find(BSTNode root, int value)
        {
            if(root == null)
            {
                return null;
            }
            if(root.val == value)
            {
                return root;
            }

            return null;
        }
    }
}
